runtime/compiler/net/Message.cpp

```diff
@@ -0,0 +1,139 @@
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "net/Message.hpp"
+#include "infra/Assert.hpp"
+#include "env/VerboseLog.hpp"
+
+namespace JITServer
+{
+   const char* const Message::DataDescriptor::_descriptorNames[] = {
+      "UINT32",
+      "INT64",
+      "UINT32",
+      "UINT64",
+      "BOOL",
+      "STRING",
+      "OBJECT",
+      "ENUM",
+      "VECTOR",
+      "SIMPLE_VECTOR",
+      "EMPTY_VECTOR",
+      "TUPLE",
+      "INVALID"
+   };
+
+uint32_t
+Message::addData(const DataDescriptor &desc, const void *dataStart, bool needs64BitAlignment)
+   {
+   // Write the descriptor itself
+   uint32_t descOffset = _buffer.writeValue(desc);
+
+   //TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "addData(dataStart=%p):  DataDescriptor(@%u): type=%d payload_size=%u dataOffset=%u, padding=%u",
+   //   dataStart, descOffset, desc.getDataType(), desc.getPayloadSize(), desc.getDataOffset(), desc.getPaddingSize());
+
+   // If the data following the descriptor needs to be 64-bit aligned,
+   // add some initial padding in the outgoing buffer and write the
+   // offset to the real payload into the descriptor
+   uint8_t initialPadding = 0;
+   if (needs64BitAlignment && !_buffer.is64BitAligned())
+      {
+      initialPadding = _buffer.alignCurrentPositionOn64Bit();
+      TR_ASSERT(initialPadding != 0, "Initial padding must be non zero because we checked alignment");
+      DataDescriptor *serializedDescriptor = _buffer.getValueAtOffset<DataDescriptor>(descOffset);
+      serializedDescriptor->addInitialPadding(initialPadding);
+      }
+
+   // Write the real data and possibly some padding at the end
+   _buffer.writeData(dataStart, desc.getPayloadSize(), desc.getPaddingSize()); 
+   _descriptorOffsets.push_back(descOffset);
+   return desc.getTotalSize() + initialPadding;
+   }
+
+void
+Message::deserialize()
+   {
+   // Assume that buffer is populated with data that defines a valid message
+   // Reconstruct the message by setting metadata and pointers to descriptors
+   // Note that the size of the entire message buffer has already been stripped
+   //
+   _buffer.readValue<MetaData>(); // This only advances curPtr in the MessageBuffer
+
+   uint32_t numDataPoints = getMetaData()->_numDataPoints;
+
+   //TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Metadata: type=%u numDataPoints=%u\n", type(), numDataPoints);
+
+   _descriptorOffsets.reserve(numDataPoints);
+   // TODO: do I need to clear the vector of _descriptorOffsets just in case?
```

Comment:
`_descriptorOffsets` is cleared in `msg.clearForWrite();` right after we send message (basically we prepare for the next message). 
```
void
CommunicationStream::writeMessage(Message &msg)
   {
   char *serialMsg = msg.serialize();
   // write serialized message to the socket
   writeBlocking(serialMsg, msg.serializedSize());
   msg.clearForWrite();
   }
```
I was afraid that it may be possible for the routine above to exit abruptly (through an exception) and the clearing to not be done. Currently the only exception explicitely thrown is
`throw JITServer::StreamFailure("JITServer I/O error: write error");` which destroys the connection, so we are safe. I don't know if `int32_t bytesWritten = (*OBIO_write)(_ssl, data + totalBytesWritten, size - totalBytesWritten);` can throw something.

In my opinion it's safer to do the clearing just before you need to do the send, but in my tests with SSL enabled I didn't find a case where the existing solution caused a problem.

Reply:
I don't have anything against clearing vector of offsets in `Message::deserialize`.
The functionality of `clearForWrite` would just need to be spread out between `writeMessage` and `deserialize` in that case.